File: algo/21.py
# Definition for singly-linked list.
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1 or list2 and list1.val > list2.val:
            list1, list2 = list2, list1
        if list1:
            list1.next = self.mergeTwoLists(list1.next, list2)
        return list1


# list1 = [1,2,4], list2 = [1,3,4]

a = ListNode(1)
a.next = ListNode(2)
a.next.next = ListNode(4)

b = ListNode(1)
b.next = ListNode(3)
b.next.next = ListNode(4)

c = Solution().mergeTwoLists(a, b)
while c:
    print(c.val)
    c = c.next

File: algo/21.kt
class ListNode(var `val`: Int) {
    var next: ListNode? = null
}

class Solution {
    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
        var l1 = list1
        var l2 = list2
        if (l1 == null || (l2 != null && l1.`val` > l2.`val`)) {
            val temp = l1
            l1 = l2
            l2 = temp
        }
        if (l1 != null) {
            l1.next = mergeTwoLists(l1.next, l2)
        }
        return l1
    }
}

var p1 = ListNode(1)
p1.next = ListNode(2)
p1.next?.next = ListNode(4)

var p2 = ListNode(1)
p2.next = ListNode(3)
p2.next?.next = ListNode(4)

var p3 = Solution().mergeTwoLists(p1, p2)
while (p3 != null) {
    println(p3?.`val`)
    p3 = p3?.next
}
